libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp

```diff
@@ -0,0 +1,325 @@
+///   Jive Compression Mode", Clemence Bouvier, Pierre Briaud, Pyrros
+///   Chaidos, Leo Perrin, Robin Salen, Vesselin Velichkov, Danny
+///   Willems, Cryptology ePrint Archive, Report 2022/840, 2019,
+///   <https://eprint.iacr.org/2022/840>
+
+namespace libsnark
+{
+
+#define FLYSTEL_DEBUG
+#define FLYSTEL_MULTIPLICATIVE_SUBGROUP_GENERATOR 7
+// alpha constant used in the Flystel E transformation. note that
+// alpha is independent of the choice of the curve, but 1/alpha (see
+// below) depends on the curve (mores specifically -- on the modulus r of
+// its scalar field Fr)
+#define FLYSTEL_ALPHA_FIVE 5
+// the mapping f(x)=x^a=y: x,y \in Fr applied in the Flystel E
+// transformation (where a is alpha) is invertible if 1/a exists. then
+// f^-1(y)=y^1/a=x. 1/a exists if gcd(a,r-1)=1 where r is the modulus
+// of Fr. 1/a can be found with the extended Euclidean algorithm which
+// finds u,v s.t. ua+v(r-1)=1 mod (r-1)=ua and so u=1/a. parameter
+// FLYSTEL_ALPHA_FIVE_INVERSE gives the value of u=1/a for a=5 for the
+// curve BLS12-381 precomputed using the Sage command
+// inverse_mod(alpha, r-1). TODO: write a function anemoi_parameters()
+// specialized by ppT that loads the precomputed constants (including
+// alpha and the multiplicative subgroup generator
+// FLYSTEL_MULTIPLICATIVE_SUBGROUP_GENERATOR) for any curve
+#define FLYSTEL_ALPHA_FIVE_INVERSE                                             \
+    "209743500700504761917790962032743863350762210002110551290414634799754324" \
+    "73805"
+
+// original constants by specification
+// for BLS12-381
+// beta = g = first multiplicative generator = 7.
+// delta = g^(-1) =
+// 14981678621464625851270783002338847382197300714436467949315331057125308909861
+// gamma = 0
+
+// constants used for debug
+#define DEBUG_FLYSTEL_ALPHA 5
+#define DEBUG_FLYSTEL_BETA 2
+#define DEBUG_FLYSTEL_GAMMA 5
+#define DEBUG_FLYSTEL_DELTA 0
+
+/// Flystel Q_gamma function for prime fields:
+/// Qf(x) = beta x^2 + gamma
+/// x: input
+/// y: output
+template<typename FieldT, size_t generator>
+class flystel_Q_gamma_prime_field_gadget : public gadget<FieldT>
+{
+private:
+    // constants
+    const FieldT beta;
+    const FieldT gamma;
+
+public:
+    // input/output
+    const linear_combination<FieldT> input;
+    const pb_variable<FieldT> output;
+
+    flystel_Q_gamma_prime_field_gadget(
+        protoboard<FieldT> &pb,
+        const linear_combination<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix);
+
+    void generate_r1cs_constraints();
+    void generate_r1cs_witness();
+};
+
+/// Flystel Q_delta function for prime fields:
+/// Qf(x) = beta x^2 + delta
+/// x: input
+/// y: output
+template<typename FieldT, size_t generator>
+class flystel_Q_delta_prime_field_gadget : public gadget<FieldT>
+{
+private:
+    // constants
+    const FieldT beta;
+    const FieldT delta;
+
+public:
+    // input/output
```

Comment:
Similarly here.  The member names give the same information.

Reply:
Fixed.